src/System.Private.CoreLib/shared/System/Number.DiyFp.cs

```diff
@@ -8,121 +8,141 @@ namespace System
 {
     internal static partial class Number
     {
-        // An exteneded floating-point data structure which is required by Grisu3 algorithm.
-        // It defines a 64-bit significand and a 32-bit exponent,
-        // which is EXTENDED compare to IEEE double precision floating-point number (53 bits significand and 11 bits exponent).
-        //
-        // Original Grisu algorithm produces suboptimal results. To shorten the output (which is part of Grisu2/Grisu3's job),
-        // we need additional 11 bits of the significand f and larger exponent e (A larger exponent range is used to avoid overflow. A 32-bit exponent is far big enough).
-        // To fully understand how Grisu3 uses this data structure to get better result, please read http://www.cs.tufts.edu/~nr/cs257/archive/florian-loitsch/printf.pdf
-        internal struct DiyFp
+        // This is a port of the `DiyFp` implementation here: https://github.com/google/double-conversion/blob/a711666ddd063eb1e4b181a6cb981d39a1fc8bac/double-conversion/diy-fp.h
+        // The backing structure and how it is used is described in more detail here: http://www.cs.tufts.edu/~nr/cs257/archive/florian-loitsch/printf.pdf
+
+        // This "Do It Yourself Floating Point" class implements a floating-point number with a ulong significand and an int exponent.
+        // Normalized DiyFp numbers will have the most significant bit of the significand set.
+        // Multiplication and Subtraction do not normalize their results.
+        // DiyFp are not designed to contain special doubles (NaN and Infinity).
+        internal readonly ref struct DiyFp
         {
-            public const int SignificandLength = 64;
+            public const int SignificandSize = 64;
 
-            // Extended significand.
-            // IEEE 754 double-precision numbers only require 53 bits significand.
-            // However, in Grisu3 we need additional 11 bits so we declare _f as ulong.
-            // Please note _f does not include sign bit.
-            private ulong _f;
+            public readonly ulong f;
```

Comment:
This avoids the overhead of the property for a perf-critical application and allows the compiler to optimize the accesses of the fields, rather than requiring us to cache things in locals ourselves.

Reply:
I get the perf value, but who is accessing this field now that requires it to be made public? With single variable names, it becomes harder to track where its going (at least when just reading the code during review).

Reply:
Nvm, I see it, for example: `Multiply(in DiyFp other)`

Reply:
The same code as before. The type is internal and I got rid of the properties (with the same name) and just made the fields available as `public readonly` instead (for the perf reasons listed above).

The `DiyFp` struct with the fields `f` and `e` is also described by the Grisu3 paper:
![image](https://user-images.githubusercontent.com/10487869/52088566-0fe17500-2561-11e9-922e-52329c40509b.png)